### ISSUE TITLE
Get room attendance list as single message

### ISSUE THREAD

**Issue by galfert** [sockethub/sockethub#277]
When asking for an attendance list of a room (via `observe`), Sockethub used to respond with a single message containing all the users of that room.

I think with introduction of the user roles this changed, Sockethub now sends separate messages for every single user. For rooms with a large number of users, this can have a significant performance impact for the client that has to process all those messages.

Is there any reason that prevents us from changing the behavior back to sending a single message?

**Comment by silverbucket** [comment]
@galfert this was something that @gregkare and I worked on explicity last year, converting each attendance entry to an object, rather than a lot string of usernames. I can't remember everything we discussed at the time but it seemed like the right thing to do there - and the old way had issues as well (i believe you might have said something about not getting detailed information about any of the users, but I can't remember exactly.

Anyone else remember any other details about this?

**Comment by galfert** [comment]
Yes, that is correct. But couldn't we just have a list of (detailed) user objects in a single message instead of strings with just the username?

**Comment by silverbucket** [comment]
@galfert are you 100% sure that the overhead is with receiving the separate payloads rather than just the fact that the app has to process all of the objects to get the user info?

**Comment by galfert** [comment]
Things we discussed during the Kosmos summit to investigate, in order to find the reason and possible solutions to the performance problem:

* remove the `console.log` statements for incoming messages in Hyperchannel
* batch up processing of incoming messages in Hyperchannel
* send attendance objects in batches from Sockethub
* send attendance list as array of activity objects from Sockethub

**Comment by raucao** [comment]
Just btw, the attendance list used in Hyperchannel atm *is* an inline list from an attendance message, but the list is just username strings.

**Comment by raucao** [comment]
Also relevant perhaps: https://www.w3.org/TR/activitystreams-core/#collections

**Comment by silverbucket** [comment]
Triage: `priority:medium`, re-scoped. XMPP now does exactly this — the attendance query response is a single message with a `members` array (`packages/platform-xmpp/src/incoming-handlers.ts`). IRC still emits one presence event per user when parsing the NAMES reply (`packages/irc2as/src/index.js`). Remaining work: make IRC match XMPP's single-message attendance shape. Related: #1110 tracks the XMPP attendance query returning empty for non-privileged users — a presence-based occupant-tracking fix there is the same mechanism IRC would need, so the two should be coordinated.
